### PR TITLE
Annotate content node endpoint with progress_fraction field.

### PR DESCRIPTION
**Pull request by jamalex**
## Summary

Annotate the `/api/content/<channel_id>/contentnode/` endpoint with a `progress_fraction` field that aggregates `ContentSummaryLog` instances corresponding to the node or its descendants, and normalizes by the total number of nodes to provide a fractional progress between 0 and 1.

Assigning @aronasorman for the API piece, and @whitzhu for the frontend piece that brings converts the number to the `progress` prop.

## TODO

- [x] Have tests been written for the new code?

### CODE REVIEW

File: kolibri/content/content_db_router.py [hunk 1]
```diff
@@ -40,18 +46,42 @@ def get_active_content_database(return_none_if_not_set=False):
...
+        # check that the content database is not empty
         if not connections[alias].introspection.table_names():
             raise KeyError("Content DB '%s' is empty!!" % alias)
 
+    # if possible, attach the default database to the content database connection to enable joins
+    attach_default_database(alias)
+
     return alias
 
+def attach_default_database(alias):
```
Comment (aronasorman) [inline kolibri/content/content_db_router.py]: Can you add a docstring explaining what this function does?

File: kolibri/content/content_db_router.py [hunk 2]
```diff
@@ -40,18 +46,42 @@ def get_active_content_database(return_none_if_not_set=False):
...
+    # if the default database uses a sqlite file, attach it to the content database connection to enable joins
+    default_db = connections.databases[DEFAULT_DB_ALIAS]
+    if default_db["ENGINE"].endswith(".sqlite3") and default_db["NAME"].endswith(".sqlite3"):
+        default_db_path = connections.databases[DEFAULT_DB_ALIAS]["NAME"]
+        try:
+            # ensure we're connected to the content database before attaching the default database
+            if not connections[alias].connection:
+                connections[alias].connect()
+            # attach the default database to the connection
+            connections[alias].connection.execute("attach database '%s' as defaultdb;" % default_db_path)
```
Comment (aronasorman) [inline kolibri/content/content_db_router.py]: Uppercase all SQL keywords!

Comment (aronasorman) [inline kolibri/content/content_db_router.py]: Can you add a comment as well, explaining what the `ATTACH` keyword does?

File: kolibri/content/content_db_router.py [hunk 3]
```diff
@@ -40,18 +46,42 @@ def get_active_content_database(return_none_if_not_set=False):
...
+            # record the fact that the default database has been attached to this content database
+            _content_databases_with_attached_default_db.add(alias)
```
Comment (aronasorman) [inline kolibri/content/content_db_router.py:87]: Can you ever remove an attached DB?

File: kolibri/content/serializers.py [hunk 4]
```diff
@@ -25,6 +28,7 @@ class ContentNodeSerializer(serializers.ModelSerializer):
     files = FileSerializer(many=True, read_only=True)
     ancestors = serializers.SerializerMethodField()
     thumbnail = serializers.SerializerMethodField()
+    progress_fraction = serializers.SerializerMethodField()
```
Comment (aronasorman) [inline kolibri/content/serializers.py:31]: Where's the corresponding migration for this?

Comment (jamalex) [inline kolibri/content/serializers.py:31]: It's a dynamic field on the serializer, it's not on a model.

File: kolibri/content/content_db_router.py (hunk 3 above)
Comment (jamalex) [inline kolibri/content/content_db_router.py:87]: You can, though here there's no need to.

The reason we do this here is to support the `default_database_is_attached` method, which gets called externally to determine whether it's attached. In the case of a non-sqlite default database, or an in-memory sqlite database, it won't attach it, and external code may need to know this and behave appropriately ([example](https://github.com/learningequality/kolibri/pull/357/files#diff-ebcc23f2084f43c36fed7b3d897da465R61)).

File: kolibri/content/content_db_router.py (hunk 2 above)
Comment (jamalex) [inline kolibri/content/content_db_router.py]: Wilco!

File: kolibri/content/content_db_router.py (hunk 1 above)
Comment (jamalex) [inline kolibri/content/content_db_router.py]: Wilco!

**Comment by jamalex** [comment]
Comments addressed.

**Comment by aronasorman** [comment]
Merge conflict.

**Pull request merged**